text/0000-autoref-for-closures.md

```diff
@@ -0,0 +1,112 @@
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Modify the `||` expression sugar so that it can expand to either `F`,
+`&F`, or `&mut F`, where `F` is a fresh struct type implementing one
+of the `Fn`/`FnMut`/`FnOnce` traits.
+
+# Motivation
+
+There are numerous reasons that one might prefer to write a function
+that takes a closure object (e.g., `&mut FnMut()`) rather than having
+that function be generic over the closure type (e.g., `F` where
+`F:FnMut()`). For example, closure objects reduce code bloat, they
+work better with object safety restrictions, and they avoid infinite
+monomorphic expansion for recursion functions.
```

Comment:
In my experience, if you can use a `&Fn(X) -> Y`  argument to avoid infinite regress in the signature for functions with recursively passed closures, you can normally use `&F where F: Fn(X) -> Y` the same way.